### DISCUSSION TITLE
Maximum heap size reached, Max Heap Size: 10485760

### DISCUSSION THREAD

**Discussion by zeroallox** [emqx/emqx#6206] [Q&A]
### Subject of the support

EMQX throws the below error and disconnects the client.

``
2021-11-11T19:07:29.905505+00:00 [error] Process:          <0.29377.0> on node 'emqx@127.0.0.1', Context:          maximum heap size reached, Max Heap Size:    10485760, Total Heap Size:  10865063, Kill:             true, Error Logger:     true, GC Info:          [{old_heap_block_size,4298223},{heap_block_size,6566731},{mbuf_size,12597},{recent_size,2222510},{stack_size,24},{old_heap_size,0},{heap_size,2972390},{bin_vheap_size,713332},{bin_vheap_block_size,1286934},{bin_old_vheap_size,0},{bin_old_vheap_block_size,75110}]
``

### Your environment
[node_dump_20211111_192516.tar.gz](https://github.com/emqx/emqx/files/7522953/node_dump_20211111_192516.tar.gz)

### LOG File
Log file has a password as it contains sensitive information. 
How can i send it securely?
[emqx-logs.txt.zip](https://github.com/emqx/emqx/files/7522952/emqx-logs.txt.zip)

**Comment by k32** [comment]
Hello,

It looks like the client got kicked since its related process consumed more memory than configured:
```
        {force_shutdown_policy,
            #{max_heap_size => 8388608,message_queue_len => 10000}},
```
Does the client subscribe to multiple high-volume topics, or a wildcard topic?

**Comment by zeroallox** [comment]
I apologize for the delay.

Yes, the client is subscribing to a wildcard topic and is expected to consume a large amount of data.

Client A -----> EMQX -----> Client B

Is it better (in general) for Client A to submit data over multiple connections instead of a single connection?

**Comment by k32** [comment]
I see. It looks like the subscriber is not being able to keep up with the number of messages it receives. EMQX is buffering messages until the total size of the buffered messages passes the configured threshold, then the client is kicked. 

If a single subscriber is not able too keep up with the message flow, you can try to run mulitple consumers in parallel using https://docs.emqx.io/en/broker/v4.3/advanced/shared-subscriptions.html#shared-subscription feature.

**Comment by zeroallox** [comment]
Understood. What about a publisher? Is there an advantage to using multiple connections when publishing a message?

**Comment by k32** [comment]
The publisher is less likely to be affected by this kind of issue.

**Comment by zeroallox** [comment]
Yes i understand. I am asking in general what the best practice is for publishing large amount of data quickly to many different topics.

**Comment by k32** [comment]
It can definitely help to publish to multiple topics instead of one, since the messages in different topics are processed in parallel. But this also means that most of the guarantees about the ordering of the messages will be lost. Depending on the kind of data you're publishing, it may or may not be acceptable.

**Comment by roysG** [comment]
Is there any way to define a bigger heap size and prevent this ugly crash?

**Comment by umeat** [comment]
Is there any update on this?

**Reply by lafirest** [reply]
@umeat 
Hi, what updates do you care about

**Comment by jsobus-neurosys** [comment]
There is a setting for force_shutdown feature, You can find details here:
https://docs.emqx.com/en/emqx/v5.0/configuration/configuration-manual.html#force-shutdown

`force_shutdown.max_heap_size` is the one You are asking about
